File: Calculator/Symbol.cpp
#include "Symbol.hpp"
#include <vector>
#include "Error.hpp"
#include <string.h>
Symbols Symbol::Get(const char* in, int& MoveSize)
{
	MoveSize = 1;
	switch (*in)
	{
	case '+':
		return Symbols::Plus;
	case '-':
		return Symbols::Minus;
	case '*':
		return Symbols::Cross;
	case '.':
		return Symbols::Point;
	case '/':
	case '\\':
		return Symbols::Slash;
	case '=':
	case '<':
	case '>':
	case '!':
		MoveSize = 2;
		if (strncmp(in, ">=", 2) == 0 || strncmp(in, "=>", 2) == 0) return Symbols::GreaterOrEqual;
		else if (strncmp(in, "<=", 2) == 0 || strncmp(in, "=<", 2) == 0) return Symbols::SmallerOrEqual;
		else if (strncmp(in, "==", 2) == 0) return Symbols::Equal;
		else if (strncmp(in, "!=", 2) == 0 || strncmp(in, "=!", 2) == 0) return Symbols::NonEqual;
		else if (MoveSize = 1, *in == '<') return Symbols::Smaller;
		else if (*in == '>') return Symbols::Greater;
		else if (*in == '=') return Symbols::Assign;
		else throw Error(Error::Codes::UnknownSymbol, in);;
	case '(':
		return Symbols::BracketOpen;
	case ')':
		return Symbols::BracketClose;
	case '^':
		return Symbols::Power;
	case ',':
		return Symbols::Comma;	
	default:
		throw Error(Error::Codes::UnknownSymbol, std::string(1, *in));
		break;
	}
}
const char* Symbol::ToString(Symbols in) {
	switch (in)
	{
	case Symbols::Void:
		return "";
	case Symbols::Plus:
		return "+";
	case Symbols::Minus:
		return "-";
	case Symbols::Cross:
		return "*";
	case Symbols::Slash:
		return "/";
	case Symbols::Equal:
		return "==";
	case Symbols::Assign: 
		return "=";
	case Symbols::Greater: 
		return ">";
	case Symbols::GreaterOrEqual: 
		return ">=";
	case Symbols::Smaller:        
		return "<";
	case Symbols::SmallerOrEqual: 
		return "<=";
	case Symbols::NonEqual:
		return "!=";
	case Symbols::BracketOpen:
		return "(";
	case Symbols::BracketClose:
		return ")";
	case Symbols::Power:       
		return "^";
	case Symbols::Comma:
		return ",";
	default:
		throw std::exception("corrupted symbol");
		break;
	}
}

File: SharpCalc/Symbol.cs
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;
namespace SharpCalc;
/// <summary>
/// used instead of <see cref="string.Substring(int, int)"/> to avoid unneccesary character copying
/// </summary>
public struct StringSegment : IComparable<StringSegment>
{
    public string Value;
    int Start;
    int End;
    public char this[int index]
    {
        get
        {
            if (index + Start > End) throw new ArgumentOutOfRangeException(nameof(index));
            return Value[index + Start];
        }
    }
    public int Length => Math.Min(Value.Length, End) - Start;
    public StringSegment(string val, int start, int end)
    {
        this.Value = val;
        this.Start = start;
        this.End = end;
    }
    public StringSegment Sub(int startindex, int length)
    {
        return new(Value, Start + startindex, Start + startindex + length);
    }
    public override string ToString()
    {
        return Value[Start..End];
    }

    public int CompareTo(StringSegment other)
    {     
        var compareIntersect = string.Compare(this.Value, this.Start, other.Value, other.Start,Math.Min(this.Length,other.Length));

        if (compareIntersect == 0)
        {
            return this.Length.CompareTo(other.Length);
        }
        else return compareIntersect;
    }

    public StringSegment(string val)
    {
        Value = val;
        Start = 0;
        End = val.Length;
    }
    public static implicit operator StringSegment(string str)
    {
        return new StringSegment(str);
    }
}
public enum Symbol
{
    Null, Plus, Minus, Cross, Slash, Equal, Assign, Greater, GreaterOrEqual, Smaller, SmallerOrEqual, NonEqual,  Power, Comma, Point, ExclamationMark,
    Invisible_FunctionCall
}
public static class SymbolIO
{
    private static readonly string[] Out =
    {
        String.Empty,"+","-","*","/","==","=",">",">=","<","<=","!=","^",",",".","!",String.Empty
    };

    private static readonly SortedDictionary<StringSegment,Symbol> In = new()
    {
        { "+",Symbol.Plus },
        { "-",Symbol.Minus },
        { "*",Symbol.Cross },
        { "/",Symbol.Slash },
        { "\\",Symbol.Slash },
        { "^",Symbol.Power },
        { ".",Symbol.Point },
        { ",",Symbol.Comma },
        { "==",Symbol.Equal },
        { "!=",Symbol.NonEqual },
        { "=!",Symbol.NonEqual },
        { "<=",Symbol.SmallerOrEqual },
        { "=<",Symbol.SmallerOrEqual },
        { ">=",Symbol.GreaterOrEqual },
        { "=>",Symbol.GreaterOrEqual },
        { "=",Symbol.Assign },
        { ">",Symbol.Greater },
        { "<",Symbol.Smaller },
        { "!",Symbol.ExclamationMark },
    };
    public static string Represent(Symbol s)
    {
        return Out[(int)s];
    }
    internal static Symbol Get(StringSegment str, out int MoveSize)
    {
        Symbol result;
        if (str.Length >= 2)
        {
            var s2 = str.Sub(0, 2);
            result = In.GetValueOrDefault(s2);
            if (result != Symbol.Null)
            {
                MoveSize = 2;
                return result;
            }
        }
        var s1 = str.Sub(0, 1);
        result = In.GetValueOrDefault(s1);
        if (result != Symbol.Null)
        {
            MoveSize = 1;
            return result;
        }
        throw new Exceptions.UnknownSymbolError(str.ToString());
    }
}